tensorflow/c/eager/dlpack.cc

```diff
@@ -0,0 +1,335 @@
+          "TF_RESOURCE is not supported by dlpack");
+      break;
+    case TF_DataType::TF_VARIANT:
+      status->status = tensorflow::errors::InvalidArgument(
+          "TF_VARIANT is not supported by dlpack");
+      break;
+    case TF_DataType::TF_UINT32:
+      dtype.code = DLDataTypeCode::kDLUInt;
+      break;
+    case TF_DataType::TF_UINT64:
+      dtype.code = DLDataTypeCode::kDLUInt;
+      break;
+    default:
+      status->status = tensorflow::errors::InvalidArgument(
+          "Unsupported TF_DataType is not supported by dlpack");
+      break;
+  }
+  return dtype;
+}
+
+DLContext  GetDLContext(TFE_TensorHandle* h, TF_Status* status) {
+  DLContext ctx;
+  const char* device_name = h->handle->DeviceName(&status->status);
+  DeviceNameUtils::ParsedName parsed_name;
+  tensorflow::DeviceNameUtils::ParseFullName(absl::string_view(device_name),
+                                             &parsed_name);
+  std::string device_type = parsed_name.type;
+  int device_id = -1;
+  if (parsed_name.has_id) {
+    device_id = parsed_name.id;
+  }  // Question? device_id?=-1
+
+  ctx.device_id = device_id;
+  if (device_type == "CPU") {
+    ctx.device_type = DLDeviceType::kDLCPU;
+  } else if (device_type == "GPU") {
+    ctx.device_type = DLDeviceType::kDLGPU;
+  } else {
+    status->status = tensorflow::errors::InvalidArgument(
+        "Unsupported Device Type for DLPack");
+  }
+
+  return ctx;
+}
+
+DLManagedTensor* TFEHandleToTFDLManagedTensorCtx(TFE_TensorHandle* h,
+                                        TF_Status* status) {
+  const Tensor* tensor = GetTensorFromHandle(h, status);
+  TF_DataType data_type = static_cast<TF_DataType>(tensor->dtype());
+  TFDLManagedTensorCtx* tfDLMTensor(new TFDLManagedTensorCtx);
+  TensorReference* tensor_ref =
+      new TensorReference(*tensor);  // This will call buf_->Ref()
+  tfDLMTensor->handle = tensor_ref;
+  tfDLMTensor->tensor.manager_ctx = tfDLMTensor;
+  tfDLMTensor->tensor.deleter = &DLManagedTensorDeleter;
+  tfDLMTensor->tensor.dl_tensor.ctx = GetDLContext(h, status);
+  int ndim = tensor->dims();
+  tfDLMTensor->tensor.dl_tensor.ndim = ndim;
+  tfDLMTensor->tensor.dl_tensor.data = TFE_TensorHandleDevicePointer(h, status);
+  tfDLMTensor->tensor.dl_tensor.dtype = GetDLDataType(data_type, status);
+
+  int64_t* shape_arr = new int64_t[ndim];
+  for (int i = 0; i < ndim; i++) {
+    shape_arr[i] = tensor->dim_size(i);
+  }
+
+  tfDLMTensor->tensor.dl_tensor.shape = shape_arr;
+
+  tfDLMTensor->tensor.dl_tensor.strides =
+      nullptr;  // Whether this is null at all the time?
+  tfDLMTensor->tensor.dl_tensor.byte_offset =
+      0;  // Whether this is 0 at all the time?
+  return &tfDLMTensor->tensor;
+}
+
+std::string FromDLContext(const DLContext& ctx, TF_Status* status) {
+  switch (ctx.device_type) {
+    case DLDeviceType::kDLCPU:
+      return "CPU:0";
+    case DLDeviceType::kDLGPU:
+      return absl::StrCat("GPU:", ctx.device_id);
+    default:
+      return "";
```

Comment:
Should we be instead returning `absl::optional<std::string>` and returning `absl::nullopt` here?